### PR TITLE
MULE-10377: Support keyless metadata fetching

### PR DESCRIPTION
**Pull request by juandesi**

### CODE REVIEW

File: modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/metadata/MetadataKeyIdObjectResolver.java [hunk 1]
```diff
@@ -51,32 +48,33 @@
...
-  }
-
-  private Object resolveMetadataKeyWhenPresent(MetadataKey key, ComponentModel componentModel) throws MetadataResolvingException {
+  public Object resolve(MetadataKey key) throws MetadataResolvingException {
 
-    final MetadataKeyIdModelProperty keyIdModelProperty = componentModel.getModelProperty(MetadataKeyIdModelProperty.class)
-        .orElseThrow(() -> buildException(format("Component '%s' doesn't have a MetadataKeyId "
-            + "parameter associated", componentModel.getName())));
+    if (keyParts.isEmpty()) {
+      return "";
```
Comment (marianogonzalez) [inline modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/metadata/MetadataKeyIdObjectResolver.java]: wouldn't an Optional be better?

File: modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/metadata/MetadataKeyIdObjectResolver.java [hunk 2]
```diff
@@ -127,69 +126,95 @@ public void visitObject(ObjectType objectType) {
...
-   * @throws MetadataResolvingException
+   * @return a new instance of the {@link MetadataKeyId} parameter {@code type}.
+   * @throws MetadataResolvingException if the Parameter type is not instantiable.
+   * @throws IllegalArgumentException   if cannot found the required default values for an specified key.
    */
-  private Object resolveMultiLevelKey(ComponentModel componentModel, MetadataKey key, Class metadataKeyType)
+  public Object resolve() throws MetadataResolvingException {
+
+    if (keyParts.isEmpty()) {
+      return "";
```
Comment (marianogonzalez) [inline modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/metadata/MetadataKeyIdObjectResolver.java]: Optional?

File: modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/metadata/MetadataMediator.java [hunk 3]
```diff
@@ -158,17 +181,14 @@ public MetadataMediator(RuntimeExtensionModel extensionModel, RuntimeComponentMo
...
-   *        {@link MetadataOutputResolver}
-   * @param key {@link MetadataKey} of the type which's structure has to be resolved, used both for input and output types
+   *                {@link MetadataOutputResolver}
+   * @param key     {@link MetadataKey} of the type which's structure has to be resolved, used both for input and output types
    * @return Successful {@link MetadataResult} if the MetadataTypes are resolved without errors Failure {@link MetadataResult}
-   *         when the Metadata retrieval of any element fails for any reason
+   * when the Metadata retrieval of any element fails for any reason
    */
-  public MetadataResult<ComponentMetadataDescriptor> getMetadata(MetadataContext context, MetadataKey key) {
+  private MetadataResult<ComponentMetadataDescriptor> getMetadata(MetadataContext context, Object key) {
```
Comment (marianogonzalez) [inline modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/metadata/MetadataMediator.java]: Optional?

File: modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/metadata/MetadataMediator.java [hunk 4]
```diff
@@ -221,33 +241,18 @@ public MetadataMediator(RuntimeExtensionModel extensionModel, RuntimeComponentMo
...
+   * @param key     {@link MetadataKey} of the type which's structure has to be resolved
    * @return Success with an {@link Optional} {@link TypeMetadataDescriptor} representing the Component's Content metadata,
-   *         resolved using the {@link MetadataContentResolver} if one is available to resolve its {@link MetadataType}, returning
-   *         {@link Optional#empty()} if no Content parameter is present Failure if the dynamic resolution fails for any reason.
+   * resolved using the {@link MetadataContentResolver} if one is available to resolve its {@link MetadataType}, returning
+   * {@link Optional#empty()} if no Content parameter is present Failure if the dynamic resolution fails for any reason.
    */
   private Optional<MetadataResult<ParameterMetadataDescriptor>> getContentMetadataDescriptor(MetadataContext context,
-                                                                                             MetadataKey key) {
+                                                                                             Object key) {
```
Comment (marianogonzalez) [inline modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/metadata/MetadataMediator.java]: Optional?

File: modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/metadata/MetadataMediator.java [hunk 5]
```diff
@@ -260,26 +265,16 @@ public MetadataMediator(RuntimeExtensionModel extensionModel, RuntimeComponentMo
...
-   * @param key {@link MetadataKey} of the type which's structure has to be resolved
+   * @param key     {@link MetadataKey} of the type which's structure has to be resolved
    * @return Success with an {@link OutputMetadataDescriptor} representing the Component's output metadata, resolved using the
-   *         {@link MetadataOutputResolver} if one is available to resolve its {@link MetadataType}. Failure if the dynamic
-   *         resolution fails for any reason.
+   * {@link MetadataOutputResolver} if one is available to resolve its {@link MetadataType}. Failure if the dynamic
+   * resolution fails for any reason.
    */
-  private MetadataResult<OutputMetadataDescriptor> getOutputMetadataDescriptor(MetadataContext context, MetadataKey key) {
+  private MetadataResult<OutputMetadataDescriptor> getOutputMetadataDescriptor(MetadataContext context, Object key) {
```
Comment (marianogonzalez) [inline modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/metadata/MetadataMediator.java]: optional

File: modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/metadata/MetadataMediator.java [hunk 6]
```diff
@@ -331,36 +326,35 @@ public MetadataMediator(RuntimeExtensionModel extensionModel, RuntimeComponentMo
    * output using the {@link MetadataOutputResolver} associated to the current component.
    *
    * @param context {@link MetadataContext} of the Metadata resolution
-   * @param key {@link MetadataKey} of the type which's structure has to be resolved
+   * @param key     {@link MetadataKey} of the type which's structure has to be resolved
    * @return a {@link MetadataResult} with the {@link MetadataType} of the component's output
    */
-  private MetadataResult<MetadataType> getOutputMetadata(final MetadataContext context, final MetadataKey key) {
+  private MetadataResult<MetadataType> getOutputMetadata(final MetadataContext context, final Object key) {
```
Comment (marianogonzalez) [inline modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/metadata/MetadataMediator.java]: optional

File: modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/metadata/MetadataMediator.java [hunk 7]
```diff
@@ -132,23 +132,46 @@ public MetadataMediator(RuntimeExtensionModel extensionModel, RuntimeComponentMo
...
-    ComponentMetadataDescriptorBuilder componentDescriptorBuilder = componentDescriptor(componentModel.getName())
-        .withParametersDescriptor(getParametersMetadataDescriptors())
-        .withOutputDescriptor(getOutputMetadataDescriptor());
-
-    Optional<MetadataResult<ParameterMetadataDescriptor>> contentDescriptor = getContentMetadataDescriptor();
-    if (contentDescriptor.isPresent()) {
-      componentDescriptorBuilder.withContentDescriptor(contentDescriptor.get());
+  public MetadataResult<ComponentMetadataDescriptor> getMetadata(MetadataContext context) {
+    try {
+      if (metadataKeyParts.size() == 0) {
```
Comment (alegmarra) [inline modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/metadata/MetadataMediator.java]: `metadataKeyParts.size() == 0` => `isKeyLess()`

File: modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/metadata/MetadataMediator.java (hunk 3 above)
Comment (alegmarra) [inline modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/metadata/MetadataMediator.java]: why? We can always return a valid result, either success or failure, but with a complete ComponentMetadataDescriptor

File: modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/metadata/MetadataKeyIdObjectResolver.java [hunk 8]
```diff
@@ -127,69 +127,95 @@ public void visitObject(ObjectType objectType) {
...
+   * @throws IllegalArgumentException   if cannot found the required default values for an specified key.
    */
-  private Object resolveMultiLevelKey(ComponentModel componentModel, MetadataKey key, Class metadataKeyType)
+  public Object resolve() throws MetadataResolvingException {
+
+    if (keyParts.isEmpty()) {
+      return NullMetadataKey.ID;
+    }
+
+    if (!keyParts.stream().allMatch(p -> p.getDefaultValue() != null)) {
```
Comment (alegmarra) [inline modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/metadata/MetadataKeyIdObjectResolver.java]: model validator

**Pull request merged**